server.go

```diff
@@ -321,41 +321,42 @@ func (srv *Server) ListenAndServe() error {
 }
 
 func (srv *Server) initServeUDP(connUDP *coapNet.ConnUDP) error {
-	srv.doneLock.Lock()
-	if srv.doneChan != nil {
-		return fmt.Errorf("server already serve connections")
+	doneChan, err := srv.initDone()
+	if err != nil {
+		return err
 	}
-	doneChan := make(chan struct{})
-	srv.doneChan = doneChan
-	srv.doneLock.Unlock()
 
 	return srv.serveUDP(newShutdownWithContext(doneChan), connUDP)
 }
 
 func (srv *Server) initServeTCP(conn *coapNet.Conn) error {
-	srv.doneLock.Lock()
-	if srv.doneChan != nil {
-		return fmt.Errorf("server already serve connections")
+	doneChan, err := srv.initDone()
+	if err != nil {
+		return err
 	}
-	doneChan := make(chan struct{})
-	srv.doneChan = doneChan
-	srv.doneLock.Unlock()
 
 	if srv.NotifyStartedFunc != nil {
 		srv.NotifyStartedFunc()
 	}
 	return srv.serveTCPConnection(newShutdownWithContext(doneChan), conn)
 }
 
-func (srv *Server) initServeDTLS(conn *coapNet.Conn) error {
+func (srv *Server) initDone() (<-chan struct{}, error) {
 	srv.doneLock.Lock()
+	defer srv.doneLock.Unlock()
 	if srv.doneChan != nil {
```

Comment:
Why to use `chan` and lock instead of the higher level [Event](https://godoc.org/google.golang.org/grpc/internal/grpcsync)?

Reply:
lock -> because shutdown can be called by any goroutines and assignation Event/channel  must be done under mutex.

Reply:
`Event` already uses a `mutex` internally.